leader/election.go

```diff
@@ -1,139 +1,204 @@
+	if err != nil {
+		return err
 	}
 
-	return false, nil
+	leaderContext, cancel := gocontext.WithCancel(ctx)
+	shutdown.AddHook(func() {
+		cancel()
+
+		// give the elector some time to release the lease
+		time.Sleep(time.Second * 2)
+	})
+
+	go elector.Run(leaderContext)
+	<-ctx.Done()
+
+	return nil
 }
 
-func createOrUpdateLease(ctx context.Context, leaseName string, attempt int) (bool, string, error) {
-	if attempt > 0 {
-		time.Sleep(time.Duration(rand.Intn(5000)) * time.Millisecond)
-	}
-	if attempt >= ctx.Properties().Int("leader.lease.attempts", 3) {
-		return false, "", fmt.Errorf("failed to acquire lease %s after %d attempts", leaseName, attempt)
+// updateLeaderLabel sets leader:true label on the current pod
+// and also removes that label from all other replicas.
+func updateLeaderLabel(ctx context.Context) {
+	backoff := retry.WithMaxRetries(3, retry.NewExponential(time.Second))
+	err := retry.Do(ctx, backoff, func(_ctx gocontext.Context) error {
+		pods, err := getAllReplicas(ctx, hostname)
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to get replicas: %w", err))
+		}
+
+		for _, pod := range pods.Items {
+			var payload string
+			if pod.Name == hostname {
+				ctx.Infof("adding leader metadata from pod: %s", pod.Name)
+				payload = `{"metadata":{"labels":{"leader":"true"}}}`
+			} else {
+				ctx.Infof("removing leader metadata from pod: %s", pod.Name)
+				payload = `{"metadata":{"labels":{"leader": null}}}`
+			}
+
+			_, err := ctx.Kubernetes().CoreV1().Pods(ctx.GetNamespace()).Patch(ctx,
+				pod.Name,
+				types.MergePatchType,
+				[]byte(payload),
+				metav1.PatchOptions{})
+			if err != nil {
+				return retry.RetryableError(err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		ctx.Errorf("failed to set label: %v", err)
 	}
-	now := metav1.MicroTime{Time: time.Now()}
-	leases := ctx.Kubernetes().CoordinationV1().Leases(ctx.GetNamespace())
-	leaseDuration := ctx.Properties().Duration("leader.lease.duration", 10*time.Minute)
-	lease, err := leases.Get(ctx, leaseName, metav1.GetOptions{})
+}
+
+// getAllReplicas returns all the pods from its parent ReplicaSet
+func getAllReplicas(ctx context.Context, thisPod string) (*corev1.PodList, error) {
+	pod, err := ctx.Kubernetes().CoreV1().Pods(ctx.GetNamespace()).Get(ctx, thisPod, metav1.GetOptions{})
 	if err != nil {
-		return false, "", err
+		return nil, err
 	}
-	if lease == nil {
-		ctx.Infof("Acquiring lease %s", leaseName)
-		lease = &v1.Lease{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      leaseName,
-				Namespace: ctx.GetNamespace(),
-			},
-			Spec: v1.LeaseSpec{
-				HolderIdentity:       &identity,
-				LeaseDurationSeconds: lo.ToPtr(int32(leaseDuration.Seconds())),
-				AcquireTime:          &now,
-				RenewTime:            &now,
-			},
-		}
-		_, err = leases.Create(ctx, lease, metav1.CreateOptions{})
-		if err != nil {
-			return false, "", err
+
+	// Get the ReplicaSet owner reference
+	var replicaSetName string
```

Comment:
This would not work for rollouts, the pods would be in different replicasets = we just need to list pods with the leader label.